substrate/utils/frame/remote-externalities/src/lib.rs

```diff
@@ -400,41 +402,135 @@ where
 			})
 	}
 
-	/// Get all the keys at `prefix` at `hash` using the paged, safe RPC methods.
-	async fn rpc_get_keys_paged(
+	/// Get keys with `prefix` at `block` in a parallel manner.
+	async fn rpc_get_keys_parallel(
 		&self,
-		prefix: StorageKey,
-		at: B::Hash,
+		prefix: &StorageKey,
+		block: B::Hash,
+		parallel: usize,
+	) -> Result<Vec<StorageKey>, &'static str> {
+		/// Divide the workload and return the start key of each chunks. Guaranteed to return a
+		/// non-empty list.
+		fn gen_start_keys(prefix: &StorageKey) -> Vec<StorageKey> {
+			let mut prefix = prefix.as_ref().to_vec();
+			let scale = 32usize.saturating_sub(prefix.len());
+
+			// no need to divide workload
+			if scale < 9 {
+				prefix.extend(vec![0; scale]);
+				return vec![StorageKey(prefix)]
+			}
+
+			// FIXME: figure out a better algo for dividing workload
+			let chunks = 16;
+			let step = 0x10000 / chunks;
+			let ext = scale - 2;
+
+			(0..chunks)
```

Comment:
Instead of dividing them up into a fixed number, we could also have a generator<->consumer pattern, where one thread creates a stream of keys that can then be consumed by a variable number of reader threads.  
Then there is no risk that one threads takes much longer than the others because of large KV pairs in a specific region.  
But anyway, if its faster for now then lets go.

Reply:
Sorry, I don't really follow, coz it seems to me there is just one role here. Or do you mean combining the processes of scraping keys and downloading the KVs? I could try it in a following PR. :))

Reply:
Yea, Producer<->Consumer pattern is a standard approach for this, but i think we should not over-optimize here; its good now, thanks!

Reply:
Right now discovering keys and downloading KVs are two separate steps. Given downloading KVs as the end goal, it does seem to me that we could cut more time by starting downloading after we have the first batch of discovered keys. Do you think it makes sense to combine the two steps? @liamaharon

The downside would be it's hard to show downloading progress, as we don't know how many keys are there until the discovery process finishes. Or is it possible to get the total amount of keys somewhere? 🤔

Reply:
@eagr yes what you described would be great! I don't think there's an api to get the total number of keys unfortunately.